File: TelerikUIExamples/TelerikUIExamples/MyPointLabelRender.m
//
//  MyPointLabelRender.m
//  TelerikUIExamples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

#import "MyPointLabelRender.h"
#import <TelerikUI/TelerikUI.h>
#import "SelectedPointLabel.h"

@implementation MyPointLabelRender
{
    SelectedPointLabel *_labelLayer;
    BOOL _isSelectedPoint;
}

- (instancetype)initWithRender:(TKChartSeriesRender *)render selectedDataIndex:(NSUInteger)dataIndex selectedSeriesIndex:(NSUInteger)seriesIndex
{
    self = [super initWithRender:render];
    if (self) {
        _selectedDataPoint = dataIndex;
        _selectedSeries = seriesIndex;
    }
    return self;
}

- (void)renderPointLabelsForSeries:(TKChartSeries *)series inRect:(CGRect)bounds context:(CGContextRef)ctx
{
    [_labelLayer removeFromSuperlayer];
    for (int i = 0; i < series.items.count; i++) {
        id<TKChartData> dataPoint = series.items[i];
        CGPoint location = [self locationForDataPoint:dataPoint forSeries:series inRect:bounds];
        if (![self isPoint:location insideRect:bounds]) {
            continue;
        }
        
        TKChartPointLabel *label = [self labelForDataPoint:dataPoint property:@"dataYValue" inSeries:series atIndex:i];
        CGRect labelRect;
        TKChartPointLabelStyle *labelStyle = series.style.pointLabelStyle;
        if (_isSelectedPoint) {
            labelRect = CGRectMake(location.x - 17.5, location.y - 10 - 2.5 * fabs(labelStyle.labelOffset.vertical), 35, 30);
            if (labelRect.origin.y < self.render.bounds.origin.y) {
                _labelLayer.isOutsideBounds = YES;
                labelRect.origin.y = location.y + 10 + 2.5 * fabs(labelStyle.labelOffset.vertical) - labelRect.size.height;
            }
            else {
                _labelLayer.isOutsideBounds = NO;
            }
            
            _labelLayer.frame = labelRect;
            [self.render addSublayer:_labelLayer];
            [_labelLayer setNeedsDisplay];
        }
        else {
            CGSize labelSize = [label sizeThatFits:bounds.size];
            labelRect = CGRectMake(location.x - labelSize.width / 2. + labelStyle.labelOffset.horizontal,
                                   location.y - labelSize.height / 2. + labelStyle.labelOffset.vertical,
                                   labelSize.width, labelSize.height);
            
            if (labelStyle.clipMode == TKChartPointLabelClipModeVisible) {
                if (labelRect.origin.y < self.render.bounds.origin.y) {
                    labelRect.origin.y = location.y - labelSize.height / 2. + fabs(labelStyle.labelOffset.vertical);
                }
            }
            
            [label drawInContext:ctx inRect:labelRect forVisualPoint:nil];
        }
    }
}

- (TKChartPointLabel *)labelForDataPoint:(id<TKChartData>)dataPoint property:(NSString *)propertyName inSeries:(TKChartSeries *)series atIndex:(NSUInteger)dataIndex
{
    if (series.index == _selectedSeries && dataIndex == _selectedDataPoint) {
        
        if (!_labelLayer) {
            _labelLayer = [[SelectedPointLabel alloc] init];
        }
        
        _labelLayer.labelStyle = series.style.pointLabelStyle;
        _labelLayer.text = [NSString stringWithFormat:@"%@", dataPoint.dataYValue];
        _isSelectedPoint = YES;
        return nil;
    }
    
    _isSelectedPoint = NO;
    return [[TKChartPointLabel alloc] initWithPoint:dataPoint series:series text:[NSString stringWithFormat:@"%@", dataPoint.dataYValue]];
}

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/MyPointLabelRender.swift
//
//  MyPointLabelRender.swift
//  TelerikUIExamplesInSwift
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import UIKit

class MyPointLabelRender: TKChartPointLabelRender {
    
    var selectedDataPointIndex: Int = 3
    var selectedSeriesIndex: Int = 0
    var labelLayer: SelectedPointLabel?
    var isSelectedPoint = false
   
    init(render: TKChartSeriesRender, selectedDataIndex: Int, selectedSeriesIndex: Int)
    {
        super.init(render: render)
        self.selectedDataPointIndex = selectedDataIndex
        self.selectedSeriesIndex = selectedSeriesIndex
    }
    
    override func renderPointLabelsForSeries(series: TKChartSeries!, inRect bounds: CGRect, context ctx: CGContext!) {
        if labelLayer != nil {
            labelLayer!.removeFromSuperlayer()
        }
        for i in 0..<series.items().count {
            let dataPoint = series.items()[i] as TKChartData
            let location = self.locationForDataPoint(dataPoint, forSeries: series, inRect: bounds)
            if !CGRectContainsPoint(bounds, location) {
                continue
            }
            
            let label: TKChartPointLabel? = self.labelForDataPoint(dataPoint, inSeries: series, atIndex:(UInt)(i))
            var labelRect = CGRectZero
            let labelStyle = series.style.pointLabelStyle
            if isSelectedPoint {
                var labelRect = CGRectMake(location.x - 17.5, location.y - 10 - 2.5 * abs(labelStyle.labelOffset.vertical), 35, 30)
                if labelRect.origin.y < bounds.origin.y {
                    labelRect.origin.y = location.y + 10 + 2.5 * abs(labelStyle.labelOffset.vertical) - labelRect.size.height
                    self.labelLayer!.isOutsideBounds = true
                }
                else {
                    self.labelLayer!.isOutsideBounds = false
                }
                
                self.labelLayer!.frame = labelRect
                self.render.addSublayer(labelLayer)
                self.labelLayer!.setNeedsDisplay()
            }
            else {
                let labelSize = label!.sizeThatFits(bounds.size)
                labelRect = CGRectMake(location.x - labelSize.width / 2.0 + labelStyle.labelOffset.horizontal,
                    location.y - labelSize.height / 2.0 + labelStyle.labelOffset.vertical,
                    labelSize.width, labelSize.height)
                
                if labelStyle!.clipMode == TKChartPointLabelClipMode.Visible {
                    if labelRect.origin.y < self.render.bounds.origin.y {
                        labelRect.origin.y = location.y - labelSize.height / 2.0 + fabs(labelStyle.labelOffset.vertical)
                    }
                }
                
                label!.drawInContext(ctx, inRect: labelRect, forVisualPoint: nil)
            }
        }
    }
    
    override func labelForDataPoint(dataPoint: TKChartData!, inSeries series: TKChartSeries!, atIndex dataIndex: UInt) -> TKChartPointLabel! {
        if Int(series.index) == self.selectedSeriesIndex && Int(dataIndex) == self.selectedDataPointIndex {
            if labelLayer == nil {
                labelLayer = SelectedPointLabel()
                labelLayer!.needsDisplayOnBoundsChange = true
                labelLayer!.contentsScale = UIScreen.mainScreen().scale
                labelLayer!.drawsAsynchronously = true
            }
            
            labelLayer!.labelStyle = series.style.pointLabelStyle
            labelLayer!.text = "\(dataPoint.dataYValue())"
            self.isSelectedPoint = true
            return nil
        }
        
        self.isSelectedPoint = false
        return TKChartPointLabel(point: dataPoint, style: series.style.pointLabelStyle, text: "\(dataPoint.dataYValue())")
    }
}
